### PR TITLE
Feature/#5 latency distribution

### PR DESCRIPTION
**Pull request by DieBauer**

### CODE REVIEW

File: src/main/scala/castalia/Delay.scala [hunk 1]
```diff
@@ -19,9 +20,143 @@ trait Delay {
...
+
+  /**
+    * Generate weibull distribution
+    *
+    * @param p95
+    * @param p99
+    * @return
+    */
+  def weibullDistribution(p95: Double, p99: Double): Distribution[Double] = {
+    val (g, b) = getWeibullParametersFromPercentiles(p95, p99)
```
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/1968742661/issues/source?bid=2998213&fileBranchId=3112713#l82)

File: src/main/scala/castalia/Delay.scala [hunk 2]
```diff
@@ -19,9 +20,143 @@ trait Delay {
...
-    dist.map(_*stdev + mean)
+  def normalDistribution(mean: Double,
+                            stdev: Double,
+                            dist: Distribution[Double]
+                           ): Distribution[Double] =
+    dist.map(_ * stdev + mean)
+
+  def gammaDistribution(k: Double, theta: Double): Distribution[Double] = {
+    Distribution.gamma(k, theta)
+  }
```
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala]: Codacy Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/scala-academy/Castalia/file/1968742661/issues/source?bid=2998213&fileBranchId=3112713#l31)

File: src/main/scala/castalia/Delay.scala [hunk 3]
```diff
@@ -19,9 +20,143 @@ trait Delay {
...
+    * continuous probability distributions defined on the interval [0, 1]
+    * parametrized by two positive shape parameters, denoted by α and β,
+    * that appear as exponents of the random variable and control the shape of the distribution.
+    *
+    * @param mu mean
+    * @param sigma stdev
+    * @return Distribution obeying beta: for use scaling should be applied
+    */
+  def betaDistribution(mu: Double, sigma: Double): Distribution[Double] = {
+    val (a, b) = getBetaParametersFromMeanAndSigma(mu, sigma)
```
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/1968742661/issues/source?bid=2998213&fileBranchId=3112713#l43)

File: src/main/java/castalia/Dist.java [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+            if (arg >= elimit) sum = 1.0 - Math.exp(arg);
+        }
+        return sum;
+    }
+
+    /** gamma quantile function */
+    public static double qgamma(double p, double alpha, double scale)
+    { /* Based on Algorithm AS 91 */
+        int MAXIT= 20;
+        double C1=0.01, C2=0.222222, C3=0.32, C4=0.4, C5=1.24,
```
Comment (TimSoethout) [inline src/main/java/castalia/Dist.java]: Codacy Issue found: [Use one line for each declaration, it enhances code readability.](https://www.codacy.com/app/scala-academy/Castalia/file/1968742700/issues/source?bid=2998213&fileBranchId=3112713#l314)

File: src/main/java/castalia/Dist.java [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+        double half = 0.5;
+        double zero = 0.0;
+        double sixten = 1.6;
+        double sqrpi = 1./SQRT2PI;
+        double thrsh = 0.66291;
+        double root32 = 5.6568542494923806;
+        final double DBL_EPSILON=2.2204460492503131e-16;
+        //final double DBL_MIN=2.2250738585072014e-308; ?
+
+        double xden, temp, xnum, result, ccum;
```
Comment (TimSoethout) [inline src/main/java/castalia/Dist.java]: Codacy Issue found: [Use one line for each declaration, it enhances code readability.](https://www.codacy.com/app/scala-academy/Castalia/file/1968742700/issues/source?bid=2998213&fileBranchId=3112713#l67)

File: src/main/java/castalia/Dist.java [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+    {
+    /* Equation 6.1.41 Abramowitz and Stegun -- extended Stirling */
+    /* See also ACM algorithm 291 */
+
+    /*	double log();*/
+        int M=6;
+        double p1[] = { 0.83333333333333101837e-1, -.277777777735865004e-2,
+                0.793650576493454e-3, -.5951896861197e-3,
+                0.83645878922e-3, -.1633436431e-2, };
+        double nfac, xsq;
```
Comment (TimSoethout) [inline src/main/java/castalia/Dist.java]: Codacy Issue found: [Use one line for each declaration, it enhances code readability.](https://www.codacy.com/app/scala-academy/Castalia/file/1968742700/issues/source?bid=2998213&fileBranchId=3112713#l203)

File: src/main/java/castalia/Dist.java [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+        s = x - 2.;
+        for (n = 0, d = 0, i = N - 1; i >= 0; i--)
+        { n = n * s + p2[i]; d = d * s + q2[i]; }
+        return (n / d);
+    }
+
+   /** gamma cdf */
+
+
+    public static double pgamma(double x, double p, double scale)
```
Comment (TimSoethout) [inline src/main/java/castalia/Dist.java]: Codacy Issue found: [The method pgamma() has an NPath complexity of 456](https://www.codacy.com/app/scala-academy/Castalia/file/1968742700/issues/source?bid=2998213&fileBranchId=3112713#l254)

File: src/main/java/castalia/Dist.java [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+
+    /** integer floor */
+    public static double fint(double x)
+    { return (x >= 0.0) ? Math.floor(x) : -Math.floor(-x); }
+
+
+    // From R source
+    /** standard normal quantile function */
+    public static double qnorms(double p)
+    { double q, r, val;
```
Comment (TimSoethout) [inline src/main/java/castalia/Dist.java]: Codacy Issue found: [Use one line for each declaration, it enhances code readability.](https://www.codacy.com/app/scala-academy/Castalia/file/1968742700/issues/source?bid=2998213&fileBranchId=3112713#l131)

File: src/main/java/castalia/Dist.java [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+        for (n = 0, d = 0, i = N - 1; i >= 0; i--)
+        { n = n * s + p2[i]; d = d * s + q2[i]; }
+        return (n / d);
+    }
+
+   /** gamma cdf */
+
+
+    public static double pgamma(double x, double p, double scale)
+    { double pn1, pn2, pn3, pn4, pn5, pn6, arg, c, rn, a, b, an;
```
Comment (TimSoethout) [inline src/main/java/castalia/Dist.java]: Codacy Issue found: [Use one line for each declaration, it enhances code readability.](https://www.codacy.com/app/scala-academy/Castalia/file/1968742700/issues/source?bid=2998213&fileBranchId=3112713#l255)

File: src/main/java/castalia/Dist.java [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+        double C1=0.01, C2=0.222222, C3=0.32, C4=0.4, C5=1.24,
+                C6=2.2, C7=4.67, C8=6.66, C9=6.73, C10=13.32,
+                C11=60.0, C12=70.0, C13=84.0, C14=105.0, C15=120.0,
+                C16=127.0, C17=140.0, C18=1175.0, C19=210.0, C20=252.0,
+                C21=2264.0, C22=294.0, C23=346.0, C24=420.0, C25=462.0,
+                C26=606.0, C27=672.0, C28=707.0, C29=735.0, C30=889.0,
+                C31=932.0, C32=966.0, C33=1141.0, C34=1182.0, C35=1278.0,
+                C36=1740.0, C37=2520.0, C38=5040.0;
+
+        double a, b, c, ch, g, p1, v;
```
Comment (TimSoethout) [inline src/main/java/castalia/Dist.java]: Codacy Issue found: [Use one line for each declaration, it enhances code readability.](https://www.codacy.com/app/scala-academy/Castalia/file/1968742700/issues/source?bid=2998213&fileBranchId=3112713#l323)

File: src/main/scala/castalia/Delay.scala [hunk 11]
```diff
@@ -19,9 +20,143 @@ trait Delay {
...
+  /**
+    * continuous probability distributions defined on the interval [0, 1]
+    * parametrized by two positive shape parameters, denoted by α and β,
+    * that appear as exponents of the random variable and control the shape of the distribution.
+    *
+    * @param mu mean
+    * @param sigma stdev
+    * @return Distribution obeying beta: for use scaling should be applied
+    */
+  def betaDistribution(mu: Double, sigma: Double): Distribution[Double] = {
```
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/1993066085/issues/source?bid=2998213&fileBranchId=3112713#l43)

File: src/main/scala/castalia/Delay.scala [hunk 12]
```diff
@@ -19,9 +20,143 @@ trait Delay {
...
+  }
+
+  /**
+    * Generate weibull distribution
+    *
+    * @param p95
+    * @param p99
+    * @return
+    */
+  def weibullDistribution(p95: Double, p99: Double): Distribution[Double] = {
```
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/1993066085/issues/source?bid=2998213&fileBranchId=3112713#l82)

File: src/main/scala/castalia/Delay.scala (hunk 11 above)
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/1993066085/issues/source?bid=2998213&fileBranchId=3112713#l43)

File: src/main/scala/castalia/Delay.scala [hunk 13]
```diff
@@ -19,9 +20,143 @@ trait Delay {
...
-}
\ No newline at end of file
+  def normalDistribution(mean: Double,
+                            stdev: Double,
+                            dist: Distribution[Double]
+                           ): Distribution[Double] =
+    dist.map(_ * stdev + mean)
+
+  def gammaDistribution(k: Double, theta: Double): Distribution[Double] = {
+    Distribution.gamma(k, theta)
```
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala]: Codacy Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/scala-academy/Castalia/file/1993066085/issues/source?bid=2998213&fileBranchId=3112713#l31)

File: src/main/scala/castalia/Delay.scala (hunk 12 above)
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/1993066085/issues/source?bid=2998213&fileBranchId=3112713#l82)

File: src/main/scala/castalia/model/Model.scala [hunk 14]
```diff
@@ -39,17 +40,44 @@ object Model extends DefaultJsonProtocol  {
...
+
+      def duration: Duration = Duration(mean.getOrElse("0 ms"))
+
+      val dist: Distribution[Double] = {
+
+        val p1 = Duration(p50.getOrElse("0 ms")).toMillis
+        val p2 = Duration(p90.getOrElse("0 ms")).toMillis
+
+        distribution match {
+          case "gamma" =>
```
Comment (TimSoethout) [inline src/main/scala/castalia/model/Model.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/1993066102/issues/source?bid=2998213&fileBranchId=3112713#l61)

File: src/main/scala/castalia/Delay.scala (hunk 13 above)
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala]: Codacy Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/scala-academy/Castalia/file/1993066085/issues/source?bid=2998213&fileBranchId=3112713#l31)

File: src/main/scala/castalia/model/Model.scala (hunk 14 above)
Comment (TimSoethout) [inline src/main/scala/castalia/model/Model.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/1993066102/issues/source?bid=2998213&fileBranchId=3112713#l61)

**Comment by TimSoethout** [comment]
Guess Codacy posted on my behalve. Do the issues make sense?

**Comment by DieBauer** [comment]
Yeah,  I've used some Java code. Loads of comments on that. That is now replaced by a library. It still says assigning to a tuple violates the naming regex, E.g. Val (a, b). codacy thinks my variable name is (a, b) instead of a and b. So I'm ignoring that one for now..

**Comment by DieBauer** [comment]
About the latencyconfig case class. Maybe it's better to have a new case class. DistributionParameters that contains the probabilities and values at those probabilities. {p1, x1, p2, x2} and let it be an Either[] with mean. Anyone thoughts on that?

File: src/main/scala/castalia/Delay.scala (hunk 11 above)
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/2016627736/issues/source?bid=2998213&fileBranchId=3112713#l43)

File: src/main/scala/castalia/Delay.scala (hunk 12 above)
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/2016627736/issues/source?bid=2998213&fileBranchId=3112713#l82)

File: src/main/scala/castalia/Delay.scala (hunk 13 above)
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala]: Codacy Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/scala-academy/Castalia/file/2016627736/issues/source?bid=2998213&fileBranchId=3112713#l31)

File: src/main/scala/castalia/model/Model.scala (hunk 14 above)
Comment (TimSoethout) [inline src/main/scala/castalia/model/Model.scala]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/2016627755/issues/source?bid=2998213&fileBranchId=3112713#l61)

File: src/main/scala/castalia/Delay.scala (hunk 11 above)
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala:42]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/2016709613/issues/source?bid=2998213&fileBranchId=3112713#l43)

File: src/main/scala/castalia/Delay.scala (hunk 12 above)
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala:81]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/2016709613/issues/source?bid=2998213&fileBranchId=3112713#l82)

File: src/main/scala/castalia/Delay.scala (hunk 13 above)
Comment (TimSoethout) [inline src/main/scala/castalia/Delay.scala:30]: Codacy Issue found: [Non ascii characters are not allowed](https://www.codacy.com/app/scala-academy/Castalia/file/2016709613/issues/source?bid=2998213&fileBranchId=3112713#l31)

File: src/main/scala/castalia/model/Model.scala (hunk 14 above)
Comment (TimSoethout) [inline src/main/scala/castalia/model/Model.scala:60]: Codacy Issue found: [Field name does not match the regular expression '^[a-z][A-Za-z0-9]*$'](https://www.codacy.com/app/scala-academy/Castalia/file/2016709632/issues/source?bid=2998213&fileBranchId=3112713#l61)

**Pull request merged**